dmoj/cptbox/ptdebug_arm64.cpp

```diff
@@ -3,109 +3,109 @@
 #include "ptbox.h"
 #include <cstdio>
 
-#ifdef HAS_DEBUGGER_ARM64
+#if defined(__arm64__) || defined(__aarch64__)
+#include <elf.h>
 #include <sys/ptrace.h>
 #include <sys/uio.h>
 
-#define ARM_x0 0
-#define ARM_x1 1
-#define ARM_x2 2
-#define ARM_x3 3
-#define ARM_x4 4
-#define ARM_x5 5
-#define ARM_x6 6
-#define ARM_x7 7
-#define ARM_x8 8
-
-#include <elf.h>
+bool pt_debugger::supports_abi(int abi) {
+    switch (abi) {
+        case PTBOX_ABI_ARM:
+        case PTBOX_ABI_ARM64:
+            return true;
+    }
+    return false;
+}
 
-void pt_debugger_arm64::pre_syscall() {
+void pt_debugger::pre_syscall() {
     struct iovec iovec;
-    iovec.iov_base = arm64_reg;
-    iovec.iov_len = sizeof arm64_reg;
+    iovec.iov_base = &regs;
+    iovec.iov_len = sizeof regs;
     if (ptrace(PTRACE_GETREGSET, tid, NT_PRSTATUS, &iovec))
         perror("ptrace(PTRACE_GETREGSET)");
 
-    arm_reg_changed = false;
+    if (iovec.iov_len == sizeof regs.arm32)
+        abi_ = PTBOX_ABI_ARM;
+    else
+        abi_ = PTBOX_ABI_ARM64;
+
+    regs_changed = false;
 }
 
-void pt_debugger_arm64::post_syscall() {
-    if (!arm_reg_changed)
+void pt_debugger::post_syscall() {
+    if (!regs_changed)
         return;
 
     struct iovec iovec;
-    iovec.iov_base = arm64_reg;
-    iovec.iov_len = sizeof arm64_reg;
+    iovec.iov_base = &regs;
+    iovec.iov_len = abi_ == PTBOX_ABI_ARM ? sizeof regs.arm32 : sizeof regs.arm64;
     if (ptrace(PTRACE_SETREGSET, tid, NT_PRSTATUS, &iovec))
         perror("ptrace(PTRACE_SETREGSET)");
 }
 
-long pt_debugger_arm64::peek_reg(int reg) {
-    return arm64_reg[reg];
-}
-
-void pt_debugger_arm64::poke_reg(int reg, long data) {
-    arm64_reg[reg] = data;
-    arm_reg_changed = true;
-}
-
-int pt_debugger_arm64::syscall() {
-    return (int) peek_reg(ARM_x8);
-}
-
-long pt_debugger_arm64::result() {
-    return peek_reg(ARM_x0);
-}
-
-void pt_debugger_arm64::result(long value) {
-    poke_reg(ARM_x0, value);
+#define INVALID_ABI(source) fprintf(stderr, source ": Invalid ABI\n")
```

Comment:
I think this should be an assert(false) and dump core, it doesn't seem like a good idea to have `syscall()` return -1 with only a warning.